commerce/domain/tax/cart/calculator/classes/TaxCartCalculatorSample.cls

```diff
@@ -34,13 +35,18 @@ public class TaxCartCalculatorSample extends CartExtension.TaxCartCalculator {
       // The cartItemCollection contains both products and shipping cart items.
       Map<String, CartExtension.CartItem> cartItemById = new Map<String, CartExtension.CartItem>();
       Map<String, CartExtension.CartItem> shippingItemById = new Map<String, CartExtension.CartItem>();
-      for (Integer i = (cartItemCollection.size() - 1); i >= 0; i--) {
-        if (cartItemCollection.get(i).getType() == CartExtension.SalesItemTypeEnum.PRODUCT) {
-          cartItemById.put(cartItemCollection.get(i).getId(), cartItemCollection.get(i));
-        } else if (cartItemCollection.get(i).getType() == CartExtension.SalesItemTypeEnum.CHARGE) {
+
+      Iterator<CartExtension.CartItem> cartItemCollectionIterator = cartItemCollection.iterator();
+
+      while (cartItemCollectionIterator.hasNext()) {
+        CartExtension.CartItem cartItem = cartItemCollectionIterator.next();
+
+        if (cartItem.getType() == CartExtension.SalesItemTypeEnum.PRODUCT) {
+          cartItemById.put(cartItem.getId(), cartItem);
```

Comment:
In CCA there is possibility for Customer to modify CCA orchestrator and invoke TaxCalculator during Cart operations (and they can be executed synchronously). In such case CartItem (or other entities in the Cart) may not have an Id yet. As a result of building the Map of Id to Object you can have several null keys and some Objects will not be available in the Map. This is an anti-pattern in CCA.

Reply:
added a null check